### PR TITLE
Update TensorRT tutorial to build-from-source.

### PR DESCRIPTION
**Pull request by KellenSunderland**
## Description ##
Update TensorRT tutorial to build-from-source.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by KellenSunderland** [comment]
FYI @Caenorst @aaronmarkham

**Comment by vandanavk** [comment]
@mxnet-label-bot add [pr-work-in-progress]

File: docs/tutorials/tensorrt/inference_with_trt.md [hunk 1]
```diff
@@ -17,29 +17,23 @@
...
-Installing MXNet with TensorRT integration is an easy process. First ensure that you are running Ubuntu 16.04, that you have updated your video drivers, and you have installed CUDA 9.0 or 9.2.  You'll need a Pascal or newer generation NVIDIA gpu.  You'll also have to download and install TensorRT libraries [instructions here](https://docs.nvidia.com/deeplearning/sdk/tensorrt-install-guide/index.html).  Once your these prerequisites installed and up-to-date you can install a special build of MXNet with TensorRT support enabled via PyPi and pip.  Install the appropriate version by running:
+Installing MXNet with TensorRT integration is an easy process. First ensure that you are running Ubuntu 18.04, and that you have updated your video drivers, and you have installed CUDA 10.0.  You'll need a Pascal or newer generation NVIDIA GPU.  You'll also have to download and install TensorRT libraries [instructions here](https://docs.nvidia.com/deeplearning/sdk/tensorrt-install-guide/index.html).  Once you have these prerequisites installed and up-to-date you can install a special build of MXNet with TensorRT support enabled via PyPi and pip.  Install the appropriate version by running:
```
Comment (aaronmarkham) [inline docs/tutorials/tensorrt/inference_with_trt.md]: What about CUDA 10.1?

File: docs/tutorials/tensorrt/inference_with_trt.md [hunk 2]
```diff
@@ -17,29 +17,23 @@
...
+pip install mxnet-tensorrt-cu10
 ```
 
-To install with CUDA 9.2:
-```
-pip install mxnet-tensorrt-cu92
-```
-
-If you are running an operating system other than Ubuntu 16.04, or just prefer to use a docker image with all prerequisites installed you can instead run:
+If you are running an operating system other than Ubuntu 18.04, or just prefer to use a docker image with all prerequisites installed you can instead run:
```
Comment (aaronmarkham) [inline docs/tutorials/tensorrt/inference_with_trt.md]: I think this is missing a pull step.

File: docs/tutorials/tensorrt/inference_with_trt.md [hunk 3]
```diff
@@ -83,26 +76,23 @@ end = time.time()
...
-
-While TensorRT integration remains experimental, we require users to set an environment variable to enable graph compilation.  You can see that at the start of this test we explicitly disabled TensorRT graph compilation support.  Next, we will run the same predictions using TensorRT.  This will require us to explicitly enable the MXNET_USE_TENSORRT environment variable, and we'll also use a slightly different API to bind our symbol.
+For this experiment we are strictly interested in inference performance, so to simplify the benchmark we'll pass a tensor filled with zeros as an input.  We then bind a symbol as usual, returning a normal MXNet executor, and we run forward on this executor in a loop.  To help improve the accuracy of our benchmarks we run a small number of predictions as a warmup before running our timed loop.  This will ensure various lazy operations, which do not represent real-world usage, have completed before we measure relative performance improvement.  On a modern PC with an RTX 2070 GPU the time taken for our MXNet baseline is **17.20s**.  Next we'll run the same model with TensorRT enabled, and see how the performance compares.
```
Comment (aaronmarkham) [inline docs/tutorials/tensorrt/inference_with_trt.md]: I think this could be simplified. Do you have to include so much detail for a toy model?
```suggestion
For this experiment we are strictly interested in inference performance, so to simplify the benchmark we'll pass a tensor filled with zeros as an input.  We will also ensure various lazy operations are excluded from the benchmark by performing a warmup before running our timed loop.  This does not represent real-world usage, but will provide a basic benchmark.  On a modern PC with an RTX 2070 GPU the time taken for our MXNet baseline is **17.20s**.  Next we'll run the same model with TensorRT enabled, and see how the performance compares.
```

**Comment by abhinavs95** [comment]
@KellenSunderland Could you have a look at the review comments? Thanks!

File: docs/tutorials/tensorrt/inference_with_trt.md (hunk 1 above)
Comment (KellenSunderland) [reply docs/tutorials/tensorrt/inference_with_trt.md]: The CI is currently targeting 10.0 but it should work with newer versions.  I'll update to say 10 or newer.

File: docs/tutorials/tensorrt/inference_with_trt.md (hunk 2 above)
Comment (KellenSunderland) [reply docs/tutorials/tensorrt/inference_with_trt.md]: Run will do a pull.

File: docs/tutorials/tensorrt/inference_with_trt.md (hunk 3 above)
Comment (KellenSunderland) [reply docs/tutorials/tensorrt/inference_with_trt.md]: Good feedback, I'll try and simplify.

**Comment by vandanavk** [comment]
@mxnet-label-bot update [pr-work-in-progress]

File: docs/tutorials/tensorrt/inference_with_trt.md [hunk 4]
```diff
@@ -83,26 +76,23 @@ end = time.time()
...
+                               grad_req='null', force_rebind=True)
+executor.copy_params_from(arg_params, aux_params)
 ```
 
-Instead of calling simple_bind directly on our symbol to return an executor, we call an experimental API from the contrib module of MXNet. This call is meant to emulate the simple_bind call, and has many of the same arguments.  One difference to note is that this call takes params in the form of a single merged dictionary to assist with a tensor cleanup pass that we'll describe below.
-
-As TensorRT integration improves our goal is to gradually deprecate this tensorrt_bind call, and allow users to use TensorRT transparently (see the [Subgraph API](https://cwiki.apache.org/confluence/display/MXNET/MXNet+Graph+Optimization+and+Quantization+based+on+subgraph+and+MKL-DNN) for more information).  When this happens, the similarity between tensorrt_bind and simple_bind should make it easy to migrate your code.
+We us a few TensorRT specific API calls from the contrib package here to setup our parameters and indicate we'd like to run inference in fp16 mode. We then call simple_bind as normal and copy our parameter dictionaries to our executor.
```
Comment (samskalicky) [inline docs/tutorials/tensorrt/inference_with_trt.md]: 'We us' ==> ?

File: docs/tutorials/tensorrt/inference_with_trt.md [hunk 5]
```diff
@@ -136,8 +126,8 @@ The examples below shows a Gluon implementation of a Wavenet before and after a
 ## After
 ![after](wavenet_optimized.svg)
 
-## Future Work
-As mentioned above, MXNet developers are excited about the possibilities of [creating APIs](https://cwiki.apache.org/confluence/display/MXNET/MXNet+Graph+Optimization+and+Quantization+based+on+subgraph+and+MKL-DNN) that deal specifically with subgraphs.  As this work matures it will bring many improvements for TensorRT users.  We hope this will also be an opportunity for other acceleration libraries to integrate with MXNet.
+## Subgraph API
+As of MXNet 1.5, MXNet developers have integrated TensorRT with MXNet via a Subgraph API.  Read more about the design of the API [here](https://cwiki.apache.org/confluence/display/MXNET/MXNet+Graph+Optimization+and+Quantization+based+on+subgraph+and+MKL-DNN).
 
 ## Thanks
 Thank you to NVIDIA for contributing this feature, and specifically thanks to Marek Kolodziej and Clement Fuji-Tsang.  Thanks to Junyuan Xie and Jun Wu for the code reviews and design feedback, and to Aaron Markham for the copy review.
```
Comment (samskalicky) [inline docs/tutorials/tensorrt/inference_with_trt.md]: 'Thank you to NVIDIA' ==> Thanks to NVIDIA

File: docs/tutorials/tensorrt/inference_with_trt.md [hunk 6]
```diff
@@ -83,26 +76,23 @@ end = time.time()
...
 ```python
 # Execute with TensorRT
 print('Building TensorRT engine')
-os.environ['MXNET_USE_TENSORRT'] = '1'
-arg_params.update(aux_params)
-all_params = dict([(k, v.as_in_context(mx.gpu(0))) for k, v in arg_params.items()])
-executor = mx.contrib.tensorrt.tensorrt_bind(sym, ctx=mx.gpu(0), all_params=all_params,
-                                             data=batch_shape, grad_req='null', force_rebind=True)
+trt_sym = sym.get_backend_symbol('TensorRT')
+mx.contrib.tensorrt.init_tensorrt_params(trt_sym, arg_params, aux_params)
```
Comment (Caenorst) [inline docs/tutorials/tensorrt/inference_with_trt.md]: The inputs `arg_params` and `aux_params` being modified by `init_tensorrt_params` is actually an unwanted behavior that I'm intending to fix, please use the returned `arg_params` / `aux_params`.

Comment (KellenSunderland) [reply docs/tutorials/tensorrt/inference_with_trt.md]: Gotcha, will do. Thanks @Caenorst

**Comment by KellenSunderland** [comment]
Should be ok to go now, would appreciate a review @aaronmarkham.

File: docs/tutorials/tensorrt/inference_with_trt.md (hunk 6 above)
Comment (KellenSunderland) [reply docs/tutorials/tensorrt/inference_with_trt.md]: Updated.

File: docs/tutorials/tensorrt/inference_with_trt.md (hunk 2 above)
Comment (aaronmarkham) [reply docs/tutorials/tensorrt/inference_with_trt.md]: I think this is still missing.

File: docs/tutorials/tensorrt/inference_with_trt.md [hunk 7]
```diff
@@ -17,29 +17,23 @@
...
+TensorRT is an inference only library, so for the purposes of this tutorial we will be using a pre-trained network, in this case a Resnet 18.  Resnets are a computationally intensive model architecture that are often used as a backbone for various computer vision tasks. Resnets are also commonly used as a reference for benchmarking deep learning library performance.  In this section we'll use a pretrained Resnet 18 from the [Gluon Model Zoo](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/model_zoo.html) and compare its inference speed with TensorRT using MXNet with TensorRT integration turned off as a baseline.
```
Comment (aaronmarkham) [inline docs/tutorials/tensorrt/inference_with_trt.md]: Can you use a relative link instead?

File: docs/tutorials/tensorrt/inference_with_trt.md [hunk 8]
```diff
@@ -118,7 +108,7 @@ for i in range(0, 10000):
 end = time.time()
 print(time.process_time() - start)
 ```
-We run timing with a warmup once more, and on the same machine, run in **18.99s**. A 1.8x speed improvement!  Speed improvements when using libraries like TensorRT can come from a variety of optimizations, but in this case our speedups are coming from a technique known as [operator fusion](http://dmlc.ml/2016/11/21/fusion-and-runtime-compilation-for-nnvm-and-tinyflow.html).
+We run timing with a warmup once more, and on the same machine, run in **9.83s**. A 1.75x speed improvement!  Speed improvements when using libraries like TensorRT can come from a variety of optimizations, but in this case our speedups are coming from a technique known as [operator fusion](http://dmlc.ml/2016/11/21/fusion-and-runtime-compilation-for-nnvm-and-tinyflow.html).
```
Comment (aaronmarkham) [inline docs/tutorials/tensorrt/inference_with_trt.md]: I think dlmc.ml is gone...

Comment (KellenSunderland) [reply docs/tutorials/tensorrt/inference_with_trt.md]: Darn, that was a good guide.  Will update.

**Comment by ptrendx** [comment]
@KellenSunderland What is the status of this PR? You listed better documentation for MXNet-TRT as 1 of the things todo for 1.6 release, which has a code freeze this week.

**Comment by KellenSunderland** [comment]
Hey @ptrendx.  Sorry given we haven't setup CD for this feature and we're behind on releases I'd recommend we remove this tutorial for the time being.  It's asking a little too much of our users to build mxnet with this feature supported correctly IMO.  I'll have a look at how CD is setup and see if we can auto-build binaries, then re-add this tutorial when the docker and python packages are up-to-date.

**Comment by ptrendx** [comment]
@aaronmarkham Could you please review this tutorial again? We would like to include it in 1.6 release.

**Comment by KellenSunderland** [comment]
Friendly ping @aaronmarkham.
